### DISCUSSION TITLE
How to set transparency of a FilamentAsset.

### DISCUSSION THREAD

**Discussion by JackKaiXing** [google/filament#4504] [Q&A]
here is the question.

Our project need fade in/out effects for glTF models loaded by gltfio, i.e. a FilamentAsset object. How can I set the transparency of a FilamentAsset object.

```C++

auto rcm = m_engine->rcm();
auto entities = m_asset->getEntities();
auto count = m_asset->getEntityCount();

for (uint32_t index = 0; index < count; index++)
{
        auto instance = m_engine->rcm()->getInstance(entities[index]);
	if (instance.isValid())
	{
		uint32_t count = rcm->getPrimitiveCount(instance);
		for (uint32_t pIndex = 0; pIndex < count; pIndex ++)
		{
			MaterialInstance* mi = rcm->getMaterialInstanceAt(instance, pIndex);
                        
                        mi->setParameter(WHAT_NAME,0.5f); // how to set transparency here?
		}
	}
}

```

**Comment by JackKaiXing** [comment]
my materialProvider is gltfio::createUberShaderLoader();

**Reply by bejado** [reply]
You can do this by setting the `baseColorFactor` material parameter.

See #1918

**Reply by romainguy** [reply]
And you need to change the blend mode too potentially.

**Reply by JackKaiXing** [reply]
@romainguy 

For the blend mode, do I have to hardcode all the glTF to be BlendMode or it could be changed at runtime.

Since, the Filament::Material object only have a `getBlendingMode()`.

**Reply by romainguy** [reply]
Because blending affects the generate shader code you indeed need to set the mode when creating the material.

**Reply by JackKaiXing** [reply]
@bejado I am working on this, but how could I get the initial `baseColorFactor`?

**Reply by JackKaiXing** [reply]
I think, the proper way is to hardcode `ubershader.mat.in`, adding a new parameter?

**Reply by bejado** [reply]
`baseColorFactor` is already defined as a parameter, you should be able to do:

```
mi->setParameter("baseColorFactor", float3(1.0f));
```

**Reply by JackKaiXing** [reply]
@bejado 

I think, `mi->setParameter("baseColorFactor", my_vector)` would overwrite the initial factor for rgb component, however alpha is the only thing I wanna to change.

**Reply by JackKaiXing** [reply]
@romainguy @bejado 

I am running into another problem about transparency.

For a single glTF file, which is consisted of several parts. I set this glTF file to be 0.5f transparency.

However, it flickers as the drawing order of each part is not fixed.

Any ideas? `setBlendOrderAt` seems to work?

**Reply by JackKaiXing** [reply]
@romainguy @bejado 

`setBlendOrderAt` works for my case. Thanks.

However, I got another question. 

How could I get the value of a parameter for current materialInstance?

**Reply by bejado** [reply]
We don't provide an API to get the current value of a parameter; it's up to the application itself to track this information.

**Reply by CallumHoughton** [reply]
I would also like to know the best way to go about doing this. I load .glb files, and the user can choose their own models to view so it isn't feasible to keep track of material parameters. Is there any other way around this?
